gravitee-apim-gateway/gravitee-apim-gateway-core/src/main/java/io/gravitee/gateway/jupiter/core/v4/endpoint/DefaultEndpointManager.java

```diff
@@ -0,0 +1,177 @@
+
+    private static final Logger log = LoggerFactory.getLogger(DefaultEndpointManager.class);
+
+    private final Api api;
+    private final EndpointConnectorPluginManager endpointConnectorPluginManager;
+    private final DeploymentContext deploymentContext;
+    private ManagedEndpointGroup defaultGroup;
+    private final Map<String, ManagedEndpointGroup> groupsByName;
+    private final Map<String, ManagedEndpoint> endpointsByName;
+
+    public DefaultEndpointManager(
+        final Api api,
+        final EndpointConnectorPluginManager endpointConnectorPluginManager,
+        final DeploymentContext deploymentContext
+    ) {
+        this.api = api;
+        this.endpointsByName = new ConcurrentHashMap<>(1);
+        this.groupsByName = new ConcurrentHashMap<>(1);
+        this.endpointConnectorPluginManager = endpointConnectorPluginManager;
+        this.deploymentContext = deploymentContext;
+    }
+
+    @Override
+    public ManagedEndpoint next() {
+        return next(EndpointCriteria.ENDPOINT_UP);
+    }
+
+    @Override
+    public ManagedEndpoint next(final EndpointCriteria criteria) {
+        final String name = criteria.getName();
+
+        if (name == null) {
+            if (defaultGroup != null && criteria.matches(defaultGroup)) {
+                return defaultGroup.next();
+            }
+        } else {
+            // First try to find an endpoint by name.
+            final ManagedEndpoint managedEndpoint = endpointsByName.get(name);
+
+            if (managedEndpoint != null) {
+                if (criteria.matches(managedEndpoint)) {
+                    return managedEndpoint;
+                }
+                return null;
+            }
+
+            final ManagedEndpointGroup managedGroup = groupsByName.get(name);
+
+            if (managedGroup != null && criteria.matches(managedGroup)) {
+                return managedGroup.next();
+            }
+        }
+        return null;
+    }
+
+    @Override
+    protected void doStart() throws Exception {
+        for (EndpointGroup endpointGroup : api.getEndpointGroups()) {
+            final ManagedEndpointGroup managedEndpointGroup = createAndStartGroup(endpointGroup);
+
+            if (defaultGroup == null) {
+                defaultGroup = managedEndpointGroup;
+            }
+        }
+    }
+
+    @Override
+    public DefaultEndpointManager preStop() {
+        for (ManagedEndpoint managedEndpoint : endpointsByName.values()) {
+            try {
+                managedEndpoint.getConnector().preStop();
+            } catch (Exception e) {
+                log.warn("An error occurred when pre-stopping endpoint connector [{}].", managedEndpoint.getDefinition().getName());
+            }
+        }
+
+        return this;
+    }
+
+    @Override
+    public void doStop() throws Exception {
+        for (String endpointName : endpointsByName.keySet()) {
+            removeEndpoint(endpointName);
```

Comment:
Should we catch exception on each call to make sure we try to remove all endpoint ?

Reply:
It is already done by the `removeEndpoint`

Reply:
No every thing is covered by the try/catch.